File: src/Creator/CreateStudent.java
package Creator;

import CourseObject.*;
import PersonObject.Advisor;
import PersonObject.Id;
import PersonObject.Password;
import PersonObject.Student;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

public class CreateStudent {
    private ArrayList<Student> students = new ArrayList<>();
    private String fileName;
    private String studentsFile;

    public CreateStudent(String fileName, String studentsFile, ArrayList<Course> courses, ArrayList<Advisor> advisors) throws JSONException, IOException{
        this.fileName = fileName;
        this.studentsFile = studentsFile;
        createStudents(courses,advisors);
    }

    private void createStudents(ArrayList<Course> courses,ArrayList<Advisor> advisors) throws IOException, JSONException {
        File allStudentFiles = new File(studentsFile);
        Scanner allStudentFilesInput = new Scanner(allStudentFiles);
        while (allStudentFilesInput.hasNextLine()) {
            try {
                String content = new String(Files.readAllBytes(Path.of(fileName + allStudentFilesInput.nextLine())));
                JSONObject jsonStudent = new JSONObject(content);
                JSONObject transcript = jsonStudent.getJSONObject("transcript");
                JSONObject registration = jsonStudent.getJSONObject("registration");

                String id = jsonStudent.getString("id");
                String name = jsonStudent.getString("name");
                String lastname = jsonStudent.getString("lastname");
                String advisorID = jsonStudent.getString("advisor");
                String booleanString = jsonStudent.getString("request");
                String[] readNotification = jsonArrToStrArr(jsonStudent.getJSONArray("readNotification"));
                String[] unreadNotification = jsonArrToStrArr(jsonStudent.getJSONArray("unreadNotification"));
                String password = jsonStudent.getString("password");

                String[] failedCoursesAr = jsonArrToStrArr(transcript.getJSONArray("failedCourses"));
                String[] completedCoursesAr = jsonArrToStrArr(transcript.getJSONArray("passedCourses"));
                String[] gradesPassed = jsonArrToStrArr(transcript.getJSONArray("gradesPassed"));
                String[] gradesFailed = jsonArrToStrArr(transcript.getJSONArray("gradesFailed"));
                int[] termPassed = jsonArrToIntArr(transcript.getJSONArray("termPassed"));

                String[] selectedCoursesAr = jsonArrToStrArr(registration.getJSONArray("selectedCourses"));
                String[] approvedCoursesAr = jsonArrToStrArr(registration.getJSONArray("approvedCourses"));

                ArrayList<GradeClass> failedCourses = setTranscriptCourses(failedCoursesAr, gradesFailed, termPassed,courses);
                ArrayList<GradeClass> passedCourses = setTranscriptCourses(completedCoursesAr, gradesPassed, termPassed,courses);

                ArrayList<Course> selectedCourses = setStudentCourses(selectedCoursesAr,courses);
                ArrayList<Course> approvedCourses = setStudentCourses(approvedCoursesAr,courses);

                int term = transcript.getInt("term");
                double gpa = calculateGPA(passedCourses, failedCourses);

                Student crtStudent = new Student(name, lastname, new Id(id), new Password(password), findAdvisor(advisorID,advisors),
                        new Transcript(gpa, term, passedCourses, failedCourses), courses);

                crtStudent.setRequest(booleanString);
                crtStudent.setReadNotifications(new ArrayList<>(Arrays.asList(readNotification)));
                crtStudent.setUnreadNotifications(new ArrayList<>(Arrays.asList(unreadNotification)));
                crtStudent.setSelectedCourses(selectedCourses);
                crtStudent.setApprovedCourses(approvedCourses);
                crtStudent.filterCourses();
                students.add(crtStudent);
            } catch (JSONException | IOException ignored) {
            }
        }
        assignStudentsToAdvisor(advisors);
        fillStudentListCourse(courses);
    }

    private String[] jsonArrToStrArr(JSONArray jsonArray) throws JSONException {
        String[] strings = new String[jsonArray.length()];
        for(int i=0; i<jsonArray.length();i++){
            strings[i]=jsonArray.getString(i);
        }
        return strings;
    }

    private int[] jsonArrToIntArr(JSONArray jsonArray) throws JSONException {
        int[] ints = new int[jsonArray.length()];
        for(int i=0; i<jsonArray.length();i++){
            ints[i]=jsonArray.getInt(i);
        }
        return ints;
    }

    private ArrayList<GradeClass> setTranscriptCourses(String[] transcriptCourses, String[] grades, int[] termPassed,ArrayList<Course> courses) {
        ArrayList<GradeClass> transcriptCourseList = new ArrayList<>();
        int j = 0;
        for (int i = 0; i < transcriptCourses.length; i++) {
            for (Course course : courses) {
                if (course.getCourseId().getId().equals(transcriptCourses[i]) && !courseExists(course, transcriptCourseList)) {
                    GradeClass gradeClass = new GradeClass(course, getCourseGrade(grades[i]));
                    if(course.getCourseType()== CourseType.NONTECHNICAL || course.getCourseName().contains("Is Sagligi ve Guvenligi")) {
                        gradeClass.setTerm(termPassed[j]);
                        j++;
                    }
                    else {
                        gradeClass.setTerm(course.getTerm());
                    }
                    transcriptCourseList.add(gradeClass);
                }
            }
        }
        return transcriptCourseList;
    }

    public ArrayList<Course> setStudentCourses(String[] studentCoursesAr, ArrayList<Course> courses) {
        ArrayList<Course> studentCoursesList = new ArrayList<>();
        for (String s : studentCoursesAr) {
            for (int j = 0; j < courses.size(); j++) {
                if ((courses.get(j) instanceof CourseSession &&
                        (courses.get(j).getCourseId().getId() + "." + ((CourseSession) courses.get(j)).getSessionId().getId()).equals(s))
                        || s.equals(courses.get(j).getCourseId().getId())) {
                    studentCoursesList.add(courses.get(j));
                }
            }
        } return studentCoursesList;
    }

    private Advisor findAdvisor(String advisorID,ArrayList<Advisor> advisors) {
        Advisor advisor = null;
        for (int i = 0; i < advisors.size(); i++) {
            if (advisors.get(i).getLecturerId().getId().equals(advisorID)) {
                advisor = advisors.get(i);
            }
        }
        return advisor;
    }

    private void assignStudentsToAdvisor(ArrayList<Advisor> advisors) {
        for (Student student : students) {
            for (Advisor advisor : advisors) {
                if (student.getAdvisor().getLecturerId().getId().equals(advisor.getLecturerId().getId())) {
                    advisor.getStudentList().add(student);
                    break;
                }
            }
        }
    }

    private void fillStudentListCourse(ArrayList<Course> courses) throws JSONException, IOException{
        String content = new String(Files.readAllBytes(Path.of("src\\JSON_Files\\courses.json")));
        JSONObject jsonObject = new JSONObject(content);
        JSONArray courseJSON = jsonObject.getJSONArray("courses");
        for(int i=0; i<courseJSON.length();i++){
            JSONObject currentCourse = courseJSON.getJSONObject(i);
            String[] courseStudentsId = jsonArrToStrArr(currentCourse.getJSONArray("studentList"));
            for (String currStudentId : courseStudentsId) {
                for (Student st : students) {
                    if (st.getStudentId().getId().equals(currStudentId) && (!(courses.get(i).getStudentList().contains(st)))) {
                        courses.get(i).getStudentList().add(st);
                        break;
                    }
                }
            }
        }
    }

    private boolean courseExists(Course course, ArrayList<GradeClass> transcriptCourseList) {
        for (GradeClass gradeClass : transcriptCourseList) {
            if (course.getCourseId().getId().equals(gradeClass.getCourse().getCourseId().getId())) {
                return true;
            }
        } return false;
    }

    public double calculateGPA(ArrayList<GradeClass> passedCourses, ArrayList<GradeClass> failedCourses) {
        double gpa=0;
        int totalCredit=0;
        for(GradeClass current: passedCourses){
            gpa += (current.getCourse().getCredit())*(letterToGrade(current.getGrade()));
            totalCredit+=current.getCourse().getCredit();
        }
        for(GradeClass current: failedCourses){
            gpa += (current.getCourse().getCredit())*(letterToGrade(current.getGrade()));
            totalCredit+=current.getCourse().getCredit();
        }
        if(totalCredit==0){
            return 0;
        }
        else{
            return gpa/totalCredit;
        }
    }

    private double letterToGrade(Grade grade){
        return switch (grade){
            case AA -> 4.0;
            case BA -> 3.5;
            case BB -> 3.0;
            case CB -> 2.5;
            case CC -> 2.0;
            case DC -> 1.5;
            case DD -> 1.0;
            case FD -> 0.5;
            case FF, DZ -> 0.0;
        };
    }

    private Grade getCourseGrade(String strGrade){
        return switch (strGrade.toUpperCase()) {
            case "AA" -> Grade.AA;
            case "BA" -> Grade.BA;
            case "BB" -> Grade.BB;
            case "CB" -> Grade.CB;
            case "CC" -> Grade.CC;
            case "DC" -> Grade.DC;
            case "DD" -> Grade.DD;
            case "FD" -> Grade.FD;
            case "FF" -> Grade.FF;
            case "DZ" -> Grade.DZ;
            default -> null;
        };
    }

    public ArrayList<Student> getStudents() {
        return students;
    }
}

File: python_development/Creator/CreateStudent.py
import json
import logging

from python_development.CourseSession import CourseSession
from python_development.CourseType import CourseType
from python_development.Grade import Grade
from python_development.GradeClass import GradeClass
from python_development.Id import Id
from python_development.Password import Password
from python_development.Student import Student
from python_development.Transcript import Transcript


class CreateStudent:
    def __init__(self, fileName, studentsFile, courses, advisors):
        self.students = []
        self.fileName = fileName
        self.studentsFile = studentsFile
        self.createStudents(courses, advisors)

    def createStudents(self, courses, advisors):
        with open(self.studentsFile, 'r') as all_student_files:
            for line in all_student_files:
                try:
                    with open(self.fileName + line.strip(), 'r') as student_file:
                        json_student = json.load(student_file)
                        transcript = json_student['transcript']
                        registration = json_student['registration']

                        studentId = json_student['id']
                        name = json_student['name']
                        lastname = json_student['lastname']

                        advisor_id = json_student['advisor']
                        boolean_string = json_student['request']
                        read_notification = json_student['readNotification']
                        unread_notification = json_student['unreadNotification']
                        password = json_student['password']

                        failed_courses_ar = transcript['failedCourses']
                        completed_courses_ar = transcript['passedCourses']
                        grades_passed = transcript['gradesPassed']
                        grades_failed = transcript['gradesFailed']
                        term_passed = transcript['termPassed']

                        selected_courses_ar = registration['selectedCourses']
                        approved_courses_ar = registration['approvedCourses']

                        failed_courses = self.setTranscriptCourses(failed_courses_ar, grades_failed,
                                                                   term_passed, courses)
                        passed_courses = self.setTranscriptCourses(completed_courses_ar, grades_passed,
                                                                   term_passed, courses)

                        selected_courses = self.setStudentCourses(selected_courses_ar, courses)
                        approved_courses = self.setStudentCourses(approved_courses_ar, courses)

                        term = transcript['term']
                        gpa = self.calculateGPA(passed_courses, failed_courses)

                        crt_student = Student(name, lastname, Id(studentId), Password(password),
                                              self.findAdvisor(advisor_id, advisors),
                                              Transcript(gpa, term, passed_courses, failed_courses), courses)

                        crt_student.request = boolean_string
                        crt_student.readNotifications = list(read_notification)
                        crt_student.unreadNotifications = list(unread_notification)
                        crt_student.selectedCourses = selected_courses
                        crt_student.approvedCourses = approved_courses
                        crt_student.filterCourses()
                        self.students.append(crt_student)
                except (json.JSONDecodeError, FileNotFoundError) as e:
                    logging.exception(f"Error in student data {line}: {e}")
                    pass

        self.assignStudentToAdvisor(advisors)
        self.fillStudentListCourse(courses)

    def setTranscriptCourses(self, transcript_courses, grades, term_passed, courses):
        transcript_course_list = []
        j = 0
        for i in range(len(transcript_courses)):
            for course in courses:
                if (course.getCourseId().getId() == transcript_courses[i] and
                        not self.courseExists(course, transcript_course_list)):
                    grade_class = GradeClass(course, self.getCourseGrade(grades[i]))
                    if (course.getCourseType() == CourseType.NONTECHNICAL or
                            "Is Sagligi ve Guvenligi" in course.getCourseName()):
                        grade_class.term = term_passed[j]
                        j += 1
                    else:
                        grade_class.term = course.getTerm()
                    transcript_course_list.append(grade_class)
        return transcript_course_list

    def setStudentCourses(self, student_courses_ar, courses):
        student_courses_list = []
        for s in student_courses_ar:
            for j in range(len(courses)):
                if isinstance(courses[j], CourseSession) and (
                        courses[j].getCourseId().getId() + "." + courses[j].getSessionId().getId() == s):
                    student_courses_list.append(courses[j])
                elif s == courses[j].getCourseId().getId():
                    student_courses_list.append(courses[j])
        return student_courses_list

    def findAdvisor(self, advisor_id, advisors):
        for advisor in advisors:
            if advisor.getPersonId().getId() == advisor_id:
                return advisor
        return None

    def assignStudentToAdvisor(self, advisors):
        for student in self.students:
            for advisor in advisors:
                if student.getAdvisor().getPersonId().getId() == advisor.getPersonId().getId():
                    advisor.getStudentList().append(student)
                    break

    def fillStudentListCourse(self, courses):
        with open("JSON_Files\\courses.json", 'r') as courses_file:
            course_json = json.load(courses_file)
            for i in range(len(course_json['courses'])):
                current_course = course_json['courses'][i]
                course_students_id = current_course['studentList']
                for curr_student_id in course_students_id:
                    for st in self.students:
                        if st.getPersonId().getId() == curr_student_id and st not in courses[i].getStudentList():
                            courses[i].getStudentList().append(st)
                            break

    def courseExists(self, course, transcript_course_list):
        for grade_class in transcript_course_list:
            if course.getCourseId().getId() == grade_class.getCourse().getCourseId().getId():
                return True
        return False

    def calculateGPA(self, passed_courses, failed_courses):
        gpa = 0
        total_credit = 0
        for current in passed_courses:
            gpa += current.course.credit * self.letterToGrade(current.grade)
            total_credit += current.course.credit
        for current in failed_courses:
            gpa += current.course.credit * self.letterToGrade(current.grade)
            total_credit += current.course.credit
        if total_credit == 0:
            return 0
        else:
            return gpa / total_credit

    def letterToGrade(self, grade):
        return {
            Grade.AA: 4.0,
            Grade.BA: 3.5,
            Grade.BB: 3.0,
            Grade.CB: 2.5,
            Grade.CC: 2.0,
            Grade.DC: 1.5,
            Grade.DD: 1.0,
            Grade.FD: 0.5,
            Grade.FF: 0.0,
            Grade.DZ: 0.0
        }[grade]

    def getCourseGrade(self, str_grade):
        return {
            "AA": Grade.AA,
            "BA": Grade.BA,
            "BB": Grade.BB,
            "CB": Grade.CB,
            "CC": Grade.CC,
            "DC": Grade.DC,
            "DD": Grade.DD,
            "FD": Grade.FD,
            "FF": Grade.FF,
            "DZ": Grade.DZ
        }[str_grade.upper()]

    def getStudents(self):
        return self.students
